### ISSUE TITLE
`bydomain` is hardcoded for `bootstrap` variance

### ISSUE THREAD

**Issue by smishr** [xKDR/Survey.jl#286]
See line 22-27 in [by.jl](https://github.com/xKDR/Survey.jl/blob/main/src/by.jl)

```julia
    for i = 1:nd
        filtered_dx = filter(!isnan, Xt_mat[i, :] .- X.statistic[i])
        push!(ses, sqrt(sum(filtered_dx .^ 2) / length(filtered_dx)))
    end
    replace!(ses, NaN => 0)
    X.SE = ses
```

This is hardcoding the variance for "bootstrap" type `ReplicateDesign`. For future additions, like `JKn`, the variance is calculated differently.

TODO: Update bydomain function with modularity for current and future methods of variance estimation.

**Comment by ayushpatnaikgit** [comment]
We need to update all the functions and think carefully. How about:

```
abstract type ReplicateDesign end 

struct Bootstrap <: ReplicateDesign
       end

struct Jackknife <: ReplicateDesign
       end
```
?

Then we can do multiple dispatch instead of if else ladders.

**Comment by smishr** [comment]
> We need to update all the functions and think carefully. How about:
> 
> ```
> abstract type ReplicateDesign end 
> 
> struct BootstrapDesign <: ReplicateDesign
>        end
> 
> struct JackknifeDesign <: ReplicateDesign
>        end
> ```
> 
> ?
> 
> Then we can do multiple dispatch instead of if else ladders.
> 
> I wish there was inheritance. We could have done BoostrapDesign <: ReplicateDesign <: SurveyDesign

Yes something like this should improve the modularity of the code.

But we need to throughly think this through. All other parts of the package should be made modular as well at the same time

**Comment by asinghvi17** [comment]
I haven't looked through the code too much, but a partial solution presents itself:

```julia
abstract type AbstractSurveyDesign end

struct SurveyDesign
...
end

abstract type AbstractReplicateDesign <: AbstractSurveyDesign end

struct BootstrapDesign <: AbstractReplicateDesign
       end

struct JackknifeDesign <: AbstractReplicateDesign
       end
```

so you basically introduce a top-level abstract type, which SurveyDesign and AbstractReplicateDesign both subtype.  Then, generic methods can be dispatched on AbstractSurveyDesign.  I've used this pattern, and seen it used, in a number of places (most notably as `Number -> (Real -> ..., Complex)`).

**Comment by asinghvi17** [comment]
Ah, I just realized I suggested two different approaches :D - let's discuss on the call if any of those make sense.
